SignalServiceKit/src/Util/DataSource.h

```diff
@@ -0,0 +1,71 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+NS_ASSUME_NONNULL_BEGIN
+
+// A protocol that abstracts away a source of NSData
+// and allows us to:
+//
+// * Lazy-load if possible.
+// * Avoid duplicate reads & writes.
+@protocol DataSource
```

Comment:
"All problems in computer science can be solved by another level of indirection"

https://en.wikipedia.org/wiki/Indirection